control-plane-operator/controllers/hostedcontrolplane/v2/kas/kms/aws.go

```diff
@@ -0,0 +1,288 @@
+	activeAWSKMSUnixSocket = fmt.Sprintf("unix://%s/%s", awsKMSVolumeMounts.Path(KasMainContainerName, kasVolumeKMSSocket().Name), activeAWSKMSUnixSocketFileName)
+)
+
+var _ IKMSProvider = &awsKMSProvider{}
+
+type awsKMSProvider struct {
+	activeKey        hyperv1.AWSKMSKeyEntry
+	backupKey        *hyperv1.AWSKMSKeyEntry
+	awsAuth          hyperv1.AWSKMSAuthSpec
+	awsRegion        string
+	kmsImage         string
+	tokenMinterImage string
+}
+
+func NewAWSKMSProvider(kmsSpec *hyperv1.AWSKMSSpec, kmsImage, tokenMinterImage string) (*awsKMSProvider, error) {
+	if kmsSpec == nil {
+		return nil, fmt.Errorf("AWS kms metadata not specified")
+	}
+	return &awsKMSProvider{
+		activeKey:        kmsSpec.ActiveKey,
+		backupKey:        kmsSpec.BackupKey,
+		awsAuth:          kmsSpec.Auth,
+		awsRegion:        kmsSpec.Region,
+		kmsImage:         kmsImage,
+		tokenMinterImage: tokenMinterImage,
+	}, nil
+}
+
+func (p *awsKMSProvider) GenerateKMSEncryptionConfig(apiVersion string) (*v1.EncryptionConfiguration, error) {
+	var providerConfiguration []v1.ProviderConfiguration
+	if len(p.activeKey.ARN) == 0 {
+		return nil, fmt.Errorf("active key metadata is nil")
+	}
+	hasher := fnv.New32()
+	_, err := hasher.Write([]byte(p.activeKey.ARN))
+	if err != nil {
+		return nil, err
+	}
+	providerConfiguration = append(providerConfiguration, v1.ProviderConfiguration{
+		KMS: &v1.KMSConfiguration{
+			APIVersion: apiVersion,
+			Name:       fmt.Sprintf("%s-%d", awsKeyNamePrefix, hasher.Sum32()),
+			Endpoint:   activeAWSKMSUnixSocket,
+			Timeout:    &metav1.Duration{Duration: 35 * time.Second},
+		},
+	})
+	if p.backupKey != nil && len(p.backupKey.ARN) > 0 {
+		hasher = fnv.New32()
+		_, err := hasher.Write([]byte(p.backupKey.ARN))
+		if err != nil {
+			return nil, err
+		}
+		providerConfiguration = append(providerConfiguration, v1.ProviderConfiguration{
+			KMS: &v1.KMSConfiguration{
+				APIVersion: apiVersion,
+				Name:       fmt.Sprintf("%s-%d", awsKeyNamePrefix, hasher.Sum32()),
+				Endpoint:   backupAWSKMSUnixSocket,
+				Timeout:    &metav1.Duration{Duration: 35 * time.Second},
+			},
+		})
+	}
+
+	if apiVersion == kmsAPIVersionV1 {
+		for _, p := range providerConfiguration {
+			p.KMS.CacheSize = ptr.To[int32](100)
+		}
+	}
+
+	providerConfiguration = append(providerConfiguration, v1.ProviderConfiguration{
+		Identity: &v1.IdentityConfiguration{},
+	})
+	encryptionConfig := &v1.EncryptionConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: v1.SchemeGroupVersion.String(),
+			Kind:       encryptionConfigurationKind,
+		},
+		Resources: []v1.ResourceConfiguration{
+			{
+				Resources: config.KMSEncryptedObjects(),
+				Providers: providerConfiguration,
+			},
+		},
+	}
+	return encryptionConfig, nil
+}
+
+func (p *awsKMSProvider) ApplyKMSConfig(podSpec *corev1.PodSpec) error {
+	if len(p.activeKey.ARN) == 0 || len(p.kmsImage) == 0 {
```

Comment:
I wonder if it'd be more consistent if we have a static podspec for this functionality that we can simply merge with the main deployment template podspec (and adapt as necessary)

Reply:
updated to let the providers return a config struct instead of taking `podSpec` as an argument.
Let me know if that looks good to you.